src/Autofac/Core/Activators/ProvidedInstance/ProvidedInstanceActivator.cs

```diff
@@ -64,23 +65,60 @@ private object GetInstance()
         /// </summary>
         public bool DisposeInstance { get; set; }
 
-        /// <summary>
-        /// Releases unmanaged and - optionally - managed resources.
-        /// </summary>
-        /// <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
+        /// <inheritdoc />
         protected override void Dispose(bool disposing)
         {
             // Only dispose of the instance here if it wasn't activated. If it was activated,
             // then either the owning lifetime scope will dispose of it automatically
             // (see InstanceLookup.Activate) or an OnRelease handler will take care of it.
-            if (disposing && DisposeInstance && _instance is IDisposable disposable && !_activated)
+            if (disposing && DisposeInstance && !_activated)
             {
-                disposable.Dispose();
+                // If we are in synchronous dispose, and an object implements IDisposable,
+                // then use it.
+                if (_instance is IDisposable disposable)
+                {
+                    disposable.Dispose();
+                }
+                else if (_instance is IAsyncDisposable)
+                {
+                    // Type only implements IAsyncDisposable, which is not valid if there
+                    // is a synchronous dispose being done.
+                    throw new InvalidOperationException(string.Format(
```

Comment:
I'm somewhat torn here. I'm curious if we should throw or just do like a `DisposeAsync().GetAwaiter().Wait()`. Or maybe nothing?

 I'm not sure about throwing during `Dispose`, like how [WCF would do if a network connection dropped](https://docs.microsoft.com/en-us/dotnet/framework/wcf/samples/use-close-abort-release-wcf-client-resources) and we ended up having to [work around that with a special `OnRelease` handler](https://github.com/autofac/Autofac.Wcf/blob/8f9a319957fc6e837d1ec4dd0fcc91eacecb9a14/src/Autofac.Integration.Wcf/RegistrationExtensions.cs#L25). I know I'm using [`ServiceBusClient`, which is only `IAsyncDisposable`](https://github.com/Azure/azure-sdk-for-net/blob/9ee228648bcd83987b67d7207b140a3e8f8af2b6/sdk/servicebus/Azure.Messaging.ServiceBus/src/Client/ServiceBusClient.cs#L20), in some of our code and an exception here would be both unexpected and possibly not easy to fix, especially if the container is getting disposed by something outside my control during app shutdown.

Reply:
So, this "throw if you're service isn't going to be disposable" was the approach we took for other service types when we added IAsyncDisposable support.

I suppose the problematic scenario is the one where you aren't implementing the component, like `ServiceBusClient`, so can't remedy the problem.

We can do the `GetAwaiter().Wait()`, it just means that user's might not be getting the nice async dispose they think they're getting.  I too am torn.

Reply:
Yeah, I didn't think the exception was going to be an issue until I hit this `ServiceBusClient` thing where it's causing all sorts of weird sync/async disposal problems because they didn't implement both. I have to imagine there are other cases like this out there, I just happen to find a concrete pain in my ass right here so I'm latching onto it.

I guess it comes down to: would it be better to dispose synchronously or not dispose at all? And if it doesn't dispose (and throws), it looks like [it stops everything else from being disposed, too.](https://github.com/autofac/Autofac/blob/79bfeeffbf753a86c7907cc58abc204602ff4014/src/Autofac/Core/Disposer.cs#L52) Which I'll admit sounded right at the time but `IAsyncDisposable` was kind of new and I didn't really _grok it_ like I'm starting to now.

It's almost like it should be a warning rather than an exception. Not that that's a thing, just like, "Hey, you did something wrong and we'll recover for you by doing it synchronously, but you're not getting the nice async you think you are."

Barring that... my _gut_ is telling me we should do what we can to dispose without throwing.

Another option might be to require folks to opt in with something like

```c#
builder.RegisterType<ServiceBusClient>().SynchronousDisposal();
```

where `SynchronousDisposal` could be an `OnRelease` thing sort of like the WCF `SafeRelease` extension where it wraps `DisposeAsync().GetAwaiter().Wait()` logic.

What's the "pit of success" here? What will make people more readily successful with the least amount of client work? Successful by default?

Reply:
Minor addition, we would need to `AsTask()` on the result of `DisposeAsync`, you cannot `GetAwaiter()` on a `ValueTask`. Only worth mentioning as another allocation that would happen during dispose.

Reply:
Like you say, raising a warning would be nice. I think the pit of success is "probably" `GetAwaiter().Wait()`, but perhaps we could write something somewhere, like to `Debug.Trace`?

Reply:
I'd go for a `Debug.WriteLine` or something similar. Maybe `Trace` instead of `Debug` - doesn't `Debug` get compiled out in release mode? I vaguely recall that being a thing. I could see possibly raising an event with the diagnostic listener thing for issues like this, too, maybe... but that'd mean propagating it around down to this level and yeahno.

Reply:
Ok, this is done; used `Trace.TraceWarning`. Attached tracers understand it's a warning level, which is good.